### PR TITLE
Add SRF module enhancements

### PR DESCRIPTION
**Pull request by lispandfound**
This PR adds a few SRF module enhancements:

1. A segment view that lets you iterate over fault segments in an SRF efficiently,
2. A `slipt` property that is the total slip at each timestep for each patch. The SRF spec stores slip in three components `slipt1`, `slipt2`, `slipt3`, but most often we care about the total slip in all three directions. 
3. A `slip` property that is the total slip for each patch. The `slip` parameter is related to `slipt` as the integral of `slipt` over time.
4. Documentation in the class docstring of each column in the points and header dataframe.

These changes are used for the plots I'm generating, and also by Cesar, who is also using the srf module for his work.

### CODE REVIEW

**Review by joelridden — APPROVED** [review]

**Comment by lispandfound** [comment]
Actually made a reasonably large change here to allow me to do the kind of at-scale processing we need to be able to do for type-5 verification. The main change is that now we actually read the SRF entirely in C code. For my testing this improves the performance of the SRF reader by something like 10x. I'm hoping this will cut down on the time to process the SRF analysis from 10 hours to 1 hour. Accordingly, the slip values are now recorded in sparse matrices that are also built in C, and so are no longer part of the points DataFrame. This is disappointing because it makes code handling slip a little more complicated, but is necessary to parse 100Mb+ SRF files efficiently. 

As for why we can't use `read_csv` and friends from pandas? Because the SRF file format has variable-width rows and weird formatting that prevents it from being read in this way. I've tried all the tricks I can think of over the past couple of weeks but I cannot see an efficient SRF parser written in python only with pandas data importing.

File: source_modelling/srf_reader.pyx [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    free(slip_matrix.col_ptr);
+    free(slip_matrix.data);
+
+cdef void read_slipt_values(FILE* srf_file, int start_column, int slip_count, sparse_matrix* slip_matrix):
+    cdef int slip_index
+    cdef void* intermediate
+    slip_matrix.row_ptr[slip_matrix.row_counter] = slip_matrix.entries
+    slip_matrix.row_counter += 1
+    if slip_count > 0:
+        # printf("%d\n", slip_count);
```
Comment (joelridden) [inline source_modelling/srf_reader.pyx:44]: Should remove these testing commented lines I reckon

**Review by claudio525 — COMMENTED** [review]
Nice work! Only comment I have is that the C-code needs a lot more comments imo. Adding both docstrings and block comments would make it significantly easier to read & work with this code.

**Comment by lispandfound** [comment]
Yeah I forgot to add comments to the srf C code

**Review by AndrewRidden-Harper — APPROVED** [review]
nice

**Review by joelridden — APPROVED** [review]

**Review by claudio525 — APPROVED** [review]

**Comment by lispandfound** [comment]
Needs to a be a draft temporarily while I figure out the SRF writing glitches.

File: source_modelling/srf.py [hunk 2]
```diff
@@ -356,7 +356,7 @@ def write_slip(srf_file: TextIO, slips: np.ndarray) -> None:
         )
 
 
-def write_srf_point(srf_file: TextIO, point: pd.Series) -> None:
+def write_srf_point(srf_file: TextIO, srf: SrfFile, point: pd.Series) -> None:
```
Comment (claudio525) [inline source_modelling/srf.py:359]: `srf` parameters is not in the docstring

File: source_modelling/srf.py [hunk 3]
```diff
@@ -366,18 +366,40 @@ def write_srf_point(srf_file: TextIO, point: pd.Series) -> None:
     point : pd.Series
         The point to write.
     """
+    index = point.name
+    slipt1 = (
```
Comment (claudio525) [inline source_modelling/srf.py:370]: Maybe worth adding a comment to clarify what is happening here?

**Review by joelridden — APPROVED** [review]

File: source_modelling/srf.py (hunk 3 above)
Comment (lispandfound) [reply source_modelling/srf.py:370]: yes, I will also change the code to use indexing that makes more sense.

Comment (lispandfound) [reply source_modelling/srf.py:370]: Actually nvm I cannot do that because of how the data is read, I'll write an explanatory comment for the indexing.

**Review by joelridden — APPROVED** [review]

**Review by sungeunbae — APPROVED** [review]
image

**Review by AndrewRidden-Harper — APPROVED** [review]
nice visual explanation in the comments

**Review by claudio525 — APPROVED** [review]

**Pull request merged**